lib/writers.js

```diff
@@ -194,63 +202,52 @@ class WriteQueue extends events.EventEmitter {
 
   run() {
     if (!this.isRunning) {
-      this.process();
+      this.isRunning = true;
+      // Use nextTick to allow the queue to build up on the current phase
+      process.nextTick(() => this.process());
     }
   }
 
   process() {
-    const self = this;
-    utils.whilst(
-      function condition() {
-        return self.queue.length > 0;
-      },
-      function whileProcess(next) {
-        self.isRunning = true;
-        const buffers = [];
-        const callbacks = [];
-        let totalLength = 0;
-        while (totalLength < self.coalescingThreshold && self.queue.length > 0) {
-          const writeItem = self.queue.shift();
-          try {
-            const data = writeItem.operation.request.write(self.encoder, writeItem.operation.streamId);
-            totalLength += data.length;
-            buffers.push(data);
-            callbacks.push(writeItem.callback);
-          }
-          catch (err) {
-            writeItem.callback(err);
-            // Break and flush what we have
-            break;
-          }
-        }
-        if (buffers.length === 0) {
-          // No need to invoke socket.write()
-          return next();
-        }
-        // Before invoking socket.write(), mark that the request has been written to avoid race conditions.
-        for (let i = 0; i < callbacks.length; i++) {
-          callbacks[i]();
-        }
-        self.netClient.write(Buffer.concat(buffers, totalLength), function socketWriteCallback(err) {
-          if (err) {
-            self.setWriteError(err);
-          }
-          // Allow IO between writes
-          setImmediate(next);
-        });
-      },
-      function loopFinished() {
-        // The queue is now empty
-        self.isRunning = false;
+    this.netClient.cork();
+    let fitsInStreamBuffer = true;
+    while (this.queue.length > 0 && fitsInStreamBuffer && !this.error) {
+      const writeItem = this.queue.shift();
+      if (!writeItem.operation.canBeWritten()) {
+        // Invoke the write callback with an error that is not going to be yielded to user
+        // as the operation has timed out or was cancelled.
+        writeItem.callback(new Error('The operation was already cancelled or timeout elapsed'));
+        continue;
+      }
+      let data;
+      try {
+        data = writeItem.operation.request.write(this.encoder, writeItem.operation.streamId);
+      }
+      catch (err) {
+        writeItem.callback(err);
+        continue;
       }
-    );
+      // Mark it as written to avoid race conditions
+      writeItem.callback();
+      fitsInStreamBuffer = this.netClient.write(data, err => this.setWriteError(err));
+    }
+    // Flush all data
+    this.netClient.uncork();
+
+    if (fitsInStreamBuffer) {
```

Comment:
I was initially thinking you would still want to call `process` or `run` here, but I realized its not necessary because if `fitsInStreamBuffer` is true, than the queue is empty or we are in an error state, so we don't need to schedule run (until more are queued).